browser/ntp_sponsored_images/new_tab_page_branded_view_counter.cc

```diff
@@ -0,0 +1,206 @@
+      base::Bind(&NewTabPageBrandedViewCounter::SetShouldShowFromPreferences,
+      base::Unretained(this)));
+  auto* ads_service_ = brave_ads::AdsServiceFactory::GetForProfile(profile);
+  if (!ads_service_) {
+    LOG(ERROR) << "Ads service is not initialized!";
+  } else {
+    is_supported_locale_ = ads_service_->IsSupportedLocale();
+  }
+}
+
+NewTabPageBrandedViewCounter::~NewTabPageBrandedViewCounter() = default;
+
+void NewTabPageBrandedViewCounter::Shutdown() {
+  NTPSponsoredImagesComponentManager* manager =
+          g_brave_browser_process->ntp_sponsored_images_component_manager();
+  if (manager->HasObserver(this))
+    manager->RemoveObserver(this);
+  auto* rewards_service_ =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile_);
+  if (rewards_service_)
+    rewards_service_->RemoveObserver(this);
+}
+
+void NewTabPageBrandedViewCounter::OnUpdated(
+    const NTPSponsoredImagesData& data) {
+  DCHECK(
+      !base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaperDemo));
+
+  // Data is updated, so change our stored data and reset any indexes.
+  // But keep view counter until branded content is seen.
+  model_.ResetCurrentWallpaperImageIndex();
+  model_.set_total_image_count(data.wallpaper_image_urls.size());
+  current_wallpaper_.reset(new NTPSponsoredImagesData(data));
+}
+
+void NewTabPageBrandedViewCounter::OnRewardsMainEnabled(
+    brave_rewards::RewardsService* rewards_service, bool rewards_main_enabled) {
+  // Reset notification state
+  ResetNotificationState();
+}
+
+void NewTabPageBrandedViewCounter::OnAdsEnabled(
+    brave_rewards::RewardsService* rewards_service, bool ads_enabled) {
+  // Reset notification state
+  ResetNotificationState();
+}
+
+void NewTabPageBrandedViewCounter::ResetNotificationState() {
+  auto* prefs = profile_->GetPrefs();
+  prefs->SetBoolean(kBrandedWallpaperNotificationDismissed, false);
+}
+
+void NewTabPageBrandedViewCounter::RegisterPageView() {
+  // Don't do any counting if we will never be showing the data
+  // since we want the count to start at the point of data being available
+  // or the user opt-in status changing.
+  if (IsBrandedWallpaperActive()) {
+    model_.RegisterPageView();
+  }
+}
+
+bool NewTabPageBrandedViewCounter::IsBrandedWallpaperActive() {
+  return (is_supported_locale_ && has_user_opted_in_ &&
+     current_wallpaper_ != nullptr);
+}
+
+bool NewTabPageBrandedViewCounter::GetShouldShowBrandedWallpaper() {
+  return IsBrandedWallpaperActive() &&
+      model_.ShouldShowBrandedWallpaper();
+}
+
+const
+NTPSponsoredImagesData& NewTabPageBrandedViewCounter::GetBrandedWallpaper() {
+  return *current_wallpaper_;
+}
+
+size_t NewTabPageBrandedViewCounter::GetWallpaperImageIndexToDisplay() {
+  return model_.current_wallpaper_image_index();
+}
+
+void NewTabPageBrandedViewCounter::SetShouldShowFromPreferences() {
+  auto* prefs = profile_->GetPrefs();
+  has_user_opted_in_ = (
```

Comment:
I think it's generally a bad idea to duplicate "storage" of prefs, especially when it isn't directly tied to a pref change observer. I think a method `IsOptedIn` or similar would be better that just checks the prefs at call time